collector/kvm_bsd.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nomeminfo
+// +build freebsd dragonfly
+
+package collector
+
+import (
+	"fmt"
+	"sync"
+)
+
+// #cgo LDFLAGS: -lkvm
+// #include "kvm_bsd.h"
+import "C"
+
+type kvm struct {
+	mu     sync.Mutex
+	hasErr bool
+}
+
+func (k *kvm) SwapUsedPages() (value uint64, err error) {
+	if k.hasErr {
```

Comment:
I think you'll also want to guard this boolean with the mutex.

Reply:
Thanks for the feedback.  I could do that.  I think my initial reasoning was that at the moment, `hasErr` is a cached thing anyway.   A second call into `_kvm_swap_used_pages` when `hasErr` is true doesn't really matter, yet.  The `_kvm_swap_used_pages` function has all the kvm state self contained, and recreates it on each call.  Perhaps the correct thing is just to rip it out, as it hasn't been proven to do anything helpful yet anyway - it's just intuition.